### ISSUE TITLE
🚀[FEA]:  [EPIC]: JAX/Warp compatibility

### ISSUE THREAD

**Issue by zubatyuk** [NVIDIA/nvalchemi-toolkit-ops#155]
### Is this a new feature, an improvement, or a change to existing functionality?

Improvement

### How would you describe the priority of this feature request

Medium

### Goal

Keep the existing nvalchemiops JAX neighbor, electrostatics, and DFT-D3 APIs working on a qualified Warp/JAX stack without changing the functions, argument layouts, or result layouts that scientific users already call.

### Why this work

Warp is removing the `warp.jax_experimental` namespace used throughout nvalchemiops and has deprecated the package's `wp.config.quiet` setting. The current dependency declaration does not state which newer Warp/JAX combination is qualified. At the same time, some JAX code treats host-side neighbor selection as though it could be traced, relies on JAX internal implementation types, or documents a DFT-D3 JIT call that does not currently work with its parameter object.

This is compatibility and correctness work, not an API redesign. It keeps the existing scientific workflows usable while making the JAX boundary honest: automatic neighbor setup is eager, fixed-shape method-specific execution is compiled, and only tested JAX behavior is documented as supported.

### Application targets

The work covers GPU-first JAX neighbor-list construction, Ewald/PME and slab electrostatics, and DFT-D3. It repairs the deprecated Warp JAX integration, makes the existing eager-versus-compiled neighbor boundary explicit, removes selected JAX-internal dependencies, and makes the documented `D3Parameters` JIT use work directly.

### Acceptance summary

After this work, electrostatics retains its current x64 behavior; `neighbor_list(...)` remains the shared eager facade; existing method-specific neighbor functions are documented and tested under `jax.jit`; and `dftd3(..., d3_params=params)` works through `jax.jit`.

### Review summary

Users do not need to change function names, positional arguments, tuple outputs, or Torch/JAX concepts. The only user-visible documentation changes explain the existing electrostatics x64 import behavior and the supported compiled neighbor call pattern. Warp's internal import migration and logging-setting migration do not add a new public API.

### Architecture

The public API remains shared where it is already shared: `neighbor_list(...)`, the existing neighbor method names and tuple/output-buffer layouts, Ewald/PME functions and parameter objects, `compute_slab_correction(...)`, `dftd3(...)`, and `D3Parameters`. Warp imports, graph modes, launch sizes, fixed-capacity buffers, and symbolic-zero handling remain private implementation details.

### Public operations

| Current condition | Target change |
| --- | --- |
| `warp-lang >= 1.13.0` has no tested range; 17 package files import `warp.jax_experimental` and one more references it in a comment. | Replace the imports with the supported top-level Warp API and qualify the dependency range. Warp 1.15 deprecates the old namespace and plans removal in 1.16.[^warp] |
| `nvalchemiops/__init__.py` sets deprecated `wp.config.quiet`. | Set `wp.config.log_level = wp.LOG_WARNING` and retain the existing package-wide `wp.init()` behavior; do not add JAX- or Torch-specific initialization. |
| Electrostatics enables `jax_enable_x64` at import time. | Preserve this behavior and its warning. Add focused documentation stating that electrostatics must be imported before creating float64 arrays or compiling float64 work.[^x64] |
| `neighbor_list(...)` selects a method from host values. | Keep the automatic facade eager. Use existing method-specific functions with fixed capacities and buffers as the compiled boundary. |
| Neighbor and electrostatics code use `jax.core.Tracer` and `jax.interpreters.ad`; `D3Parameters` is not a PyTree. | Replace supported internals where possible, isolate only a necessary symbolic-zero seam, and register `D3Parameters` as a PyTree. |

### Warp/JAX boundary

```python
# Before
from warp.jax_experimental import GraphMode, jax_callable, jax_kernel

# After
from warp import JaxCallableGraphMode, jax_callable, jax_kernel

# GraphMode.WARP becomes JaxCallableGraphMode.WARP.
```

### Package-wide Warp initialization

`nvalchemiops/__init__.py` runs before either `nvalchemiops.jax` or `nvalchemiops.torch` and is the only production location that calls `wp.init()`. Both backends share that initialized Warp runtime. This proposal keeps that ownership unchanged; it changes only the deprecated logging setting.

```python
# nvalchemiops/__init__.py: shared by Torch and JAX
wp.config.log_level = wp.LOG_WARNING
wp.init()
```

### JAX execution contract

```python
# Eager setup chooses an existing algorithm and capacity.
method = suggest_neighbor_list_method(batch_ptr, cell, pbc, cutoff, batch_idx=batch_idx)

# Compiled execution uses the selected method-specific entry point and fixed-shape buffers.
neighbors = jax.jit(naive_neighbor_list)(positions, cutoff, cell=cell, pbc=pbc, max_neighbors=max_neighbors, neighbor_matrix=neighbor_matrix, num_neighbors=num_neighbors)
```

## 3. JAX execution flow and required data operations

The automatic `neighbor_list(...)` call may inspect host values, choose a method, allocate output buffers, and retry after overflow. It therefore stays outside `jax.jit`. A compiled workflow performs that setup once, then calls the existing method-specific neighbor function with an explicit capacity and supplied output buffers. The existing count and overflow conventions remain the observable result.

Electrostatics keeps its import-time x64 guard. The documentation change explains the current ordering requirement but does not ask callers to configure JAX themselves. `wp.init()` remains a single package-root action shared by Torch and JAX: both backends import and execute Warp code, and neither has a separate initialization path. Moving or duplicating it would only add backend-specific behavior without a demonstrated user benefit.

## 4. Toolkit high-level API and migration

| Area | Public compatibility promise | Migration note |
| --- | --- | --- |
| Warp JAX integration | Preserve nvalchemiops signatures, aliases, graph-mode behavior, preload device, output order, and derivative flags. | Internal import migration only. |
| Package initialization | Keep one package-root `wp.init()` and its current targeted initialization error for both Torch and JAX. | Replace the deprecated `wp.config.quiet = True` with `wp.config.log_level = wp.LOG_WARNING`; do not add backend-specific initialization. |
| Electrostatics x64 | Preserve import-time enabling and the warning emitted when it changes the global JAX setting. | Add documentation only; no code or caller migration. |
| Neighbor compilation | Preserve `neighbor_list(...)` as host/eager and retain existing method-specific functions and tuple layouts. | Document fixed-capacity, explicit-buffer calls as the compiled boundary. |
| Electrostatics and DFT-D3 | Preserve names, accepted positional calls, aliases, parameter objects, and tuple outputs. | `D3Parameters` gains JAX PyTree behavior without a new public function. |

```python
# Existing public DFT-D3 call after D3Parameters registration.
jit_dftd3 = jax.jit(partial(dftd3, a1=a1, a2=a2, s8=s8, num_systems=num_systems))
energy, forces, coord_num = jit_dftd3(positions=positions, numbers=numbers, neighbor_list=neighbor_list, neighbor_ptr=neighbor_ptr, d3_params=params)
```

### Warp registration and package setup

`pyproject.toml` and the 17 modules under `nvalchemiops/jax/` that import `warp.jax_experimental` move to the supported top-level imports. `nvalchemiops/math/spline.py` loses its stale experimental-namespace comment. `nvalchemiops/__init__.py` replaces `wp.config.quiet = True` with `wp.config.log_level = wp.LOG_WARNING` and retains its single shared `wp.init()` call.

### Neighbor compilation seam

`nvalchemiops/jax/neighbors/__init__.py` and `_dispatch.py` retain host-side method selection. `cluster_tile.py` and `batch_cluster_tile.py` remove or narrow their `jax.core.Tracer` type checks so the affected paths use explicit capacity and metadata or raise an actionable unsupported-trace error. No JAX-only plan, state, or result object is introduced.

### Electrostatics derivative seam

`ewald.py`, `pme.py`, and `slab.py` currently use `jax.interpreters.ad.Zero` in custom derivative rules. The change replaces that internal type check with a supported JAX mechanism when the qualified version range provides one. If it does not, one small compatibility predicate remains adjacent to those derivative rules and is version-tested.

### DFT-D3 parameter object

`_dftd3.py` registers the existing dataclass with four dynamic array leaves and one static metadata field.

```python
jax.tree_util.register_dataclass(D3Parameters, data_fields=["rcov", "r4r2", "c6ab", "cn_ref"], meta_fields=["interp_mesh"])
```

## 6. Benchmarks and acceptance gates

| Gate | Evidence required | Pass condition |
| --- | --- | --- |
| Public compatibility | Project-native tests cover preserved calls, layouts, aliases, positional behavior, and x64 import behavior. | Existing supported calls still work with unchanged observable results. |
| Numerical correctness | An independent reference, invariant, or finite difference is selected before changing the numerical path. | Results meet the declared tolerance on risk-relevant inputs. |
| JAX execution | Focused eager, JIT, and only the explicitly claimed derivative tests. | No tracer concretization and no transform claim beyond the tested boundary. |
| Native CUDA | Recorded CUDA runtime and driver, JAX/JAXLIB/plugin, Warp, device, and x64 state. | The intended Warp-backed path executed; a reference-only pass is not enough. |

## 7. Risks and open questions

The automatic neighbor dispatcher should remain eager because it chooses an algorithm from values and may allocate or retry. Users who need compilation call an existing method-specific function after eager setup. Making automatic neighbor selection generally traceable would be separate work.

The exact Warp/JAX/JAXLIB bounds are not guessed here. Clean-install and focused-transform results from PR 01 determine whether an upper Warp bound is needed. If JAX has no supported symbolic-zero replacement in that qualified range, PR 03 keeps one small, tested predicate at the derivative call sites.

## 8. Dependencies and testing

PR 01 qualifies the Warp, JAX, JAXLIB, CUDA, and driver combination before later PRs rely on it. Each PR records the accepted public calls and errors it touches in its test plan before editing; this is part of its review rather than a separate project phase. The inventory covers positional arguments, aliases, tuple and output-buffer layouts, dtype behavior, and relevant eager/JIT/derivative behavior.

PR 02 uses all-pairs or gather/scatter checks for changed neighbor paths. PR 03 retains Ewald/PME identity or independent-reference checks and slab's analytical-reference and first/nested-gradient tests. PR 04 uses direct JIT invocation of `D3Parameters`, existing matrix/list and empty-system coverage, and finite-difference or independent position-force checks.
